src/main/java/org/mockbukkit/mockbukkit/block/state/BlockStateMock.java

```diff
@@ -334,8 +337,27 @@ public boolean isCollidable()
 	@Override
 	public @NotNull BlockState copy()
 	{
-		//TODO Auto-generated method stub
-		throw new UnimplementedOperationException();
+		Class<? extends BlockState> target = this.getClass();
+		try
+		{
+			for (Constructor<?> constructor : target.getDeclaredConstructors())
+			{
+				// This will make sure we find the most suitable constructor for this
+				if (constructor.getParameterCount() == 1
+						&& constructor.getParameterTypes()[0].isAssignableFrom(target))
+				{
+					return (BlockState) constructor.newInstance(this);
+				}
+			}
+
+			throw new NoSuchMethodException(
+					"Cannot find an BlockState constructor for the class \"" + target.getName() + "\"");
+		}
+		catch (SecurityException | InstantiationException | IllegalAccessException | InvocationTargetException
+			   | NoSuchMethodException e)
+		{
+			throw new BlockStateInitException(e);
+		}
```

Comment:
Personally, I'm not a big fan of using reflection here. I would suggest manually calling the constructor, since it makes it easier to identify usages:
```java
return new BlockStateMock(this);
```

Reply:
This creates one of the specific state mocks
- BannerStateMock
- BarrelStateMock
- BlastFurnaceStateMock
- BrewingStandStateMock
- ChestStateMock
- DispenserStateMock
- DropperStateMock
- FurnaceStateMock
- HopperStateMock
- LecternStateMock
- ShulkerBoxStateMock
- SmokerStateMock

Reply:
Yes, but personally i would prefer to use a [Prototype Constructor](https://refactoring.guru/design-patterns/prototype) to create the items instead of using reflection. Basically something similar to what was done in  #1224.

This is my personal opinion, but i would like to know your opinions on this topic: @ens-gijs , @Thorinwasher and/or @thelooter

Reply:
From my point of view it's fine. Probably a bit harder to comprehend, but that comes with more automation

Reply:
implemented non-abstract, non-reflection, solution.